### PR TITLE
Protection against invalid pointers in UniqueMerger added.

### PR DESCRIPTION
**Pull request by laurenhay**
#### PR description:

Second attempt at fixing issue reported  here #30112.  First attempt was #30114 by @rappoccio , but the suggestion by @gpetruc was to move the protection upstream.

#### PR validation:

Ran test on a specific event that was known to throw the error.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: CommonTools/UtilAlgos/interface/UniqueMerger.h [hunk 1]
```diff
@@ -58,14 +61,18 @@ UniqueMerger<InputCollection, OutputCollection, P>::~UniqueMerger() {}
...
   set_type coll_set;
   for (typename vtoken::const_iterator s = srcToken_.begin(); s != srcToken_.end(); ++s) {
     edm::Handle<InputCollection> h;
     evt.getByToken(*s, h);
     for (typename InputCollection::const_iterator c = h->begin(); c != h->end(); ++c) {
-      coll_set.emplace(P::clone(*c));
+      if (P::clone(*c).isNonnull() && P::clone(*c).isAvailable()) {
+        coll_set.emplace(P::clone(*c));
+      } else {
+        edm::LogWarning("InvalidPointer") << "Found an invalid pointer. Will not merge to collection." << std::endl;
```
[the remark this answers is not in the document]
Comment (slava77) [reply CommonTools/UtilAlgos/interface/UniqueMerger.h:74]: According to https://github.com/cms-sw/cmssw/issues/30112#issuecomment-641435882, this is an expected behavior.
We do not issue warnings for expected behavior.
However, for a generic merger this behavior should be configurable.
Please add a config option to skip null/unavailable pointers.

File: CommonTools/UtilAlgos/interface/UniqueMerger.h [hunk 2]
```diff
@@ -58,14 +61,18 @@ UniqueMerger<InputCollection, OutputCollection, P>::~UniqueMerger() {}
...
 void UniqueMerger<InputCollection, OutputCollection, P>::produce(edm::StreamID,
-                                                              edm::Event& evt,
-                                                              const edm::EventSetup&) const {
+                                                                 edm::Event& evt,
+                                                                 const edm::EventSetup&) const {
   set_type coll_set;
   for (typename vtoken::const_iterator s = srcToken_.begin(); s != srcToken_.end(); ++s) {
     edm::Handle<InputCollection> h;
     evt.getByToken(*s, h);
     for (typename InputCollection::const_iterator c = h->begin(); c != h->end(); ++c) {
```
[the remark this answers is not in the document]
Comment (rappoccio) [reply CommonTools/UtilAlgos/interface/UniqueMerger.h:70]: 1. The `get` method doesn't return an iterable, it returns bool. 
1. Also doing range-based looping over `h` in this case indeed does some template tricks when `InputCollection` is a `vector<Ptr<T>>`. 

We've changed the previous loop over `src_Token_` to be range-based, though.

File: CommonTools/UtilAlgos/interface/UniqueMerger.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (rappoccio) [reply CommonTools/UtilAlgos/interface/UniqueMerger.h:74]: We're a bit confused why anyone would write a null / unavailable pointer on purpose? Do you mean we issue warnings only if configured?

File: CommonTools/UtilAlgos/interface/UniqueMerger.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (slava77) [reply CommonTools/UtilAlgos/interface/UniqueMerger.h]: get returns the product const reference
https://github.com/cms-sw/cmssw/blob/master/FWCore/Framework/interface/Event.h#L192

File: CommonTools/UtilAlgos/interface/UniqueMerger.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (slava77) [reply CommonTools/UtilAlgos/interface/UniqueMerger.h]: a generic merger should have a generic or flexible behavior.
So,
- a flag to skip nulls
- a flag to issue warning on skip

File: CommonTools/UtilAlgos/interface/UniqueMerger.h [hunk 3]
```diff
@@ -14,16 +15,19 @@
...
  */
+
+#include <typeinfo>
 #include "FWCore/Framework/interface/global/EDProducer.h"
 #include "FWCore/Framework/interface/Event.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/Utilities/interface/transform.h"
 #include "FWCore/Utilities/interface/InputTag.h"
 #include "DataFormats/Common/interface/CloneTrait.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
```
Comment (perrotta) [inline CommonTools/UtilAlgos/interface/UniqueMerger.h:30]: This include is probably not needed any more, now

File: CommonTools/UtilAlgos/interface/UniqueMerger.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (rappoccio) [reply CommonTools/UtilAlgos/interface/UniqueMerger.h]: Ahh, yes, okay, we got that working. Thanks.

File: CommonTools/UtilAlgos/interface/UniqueMerger.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (rappoccio) [reply CommonTools/UtilAlgos/interface/UniqueMerger.h]: OK. Implementation incoming.

File: CommonTools/UtilAlgos/interface/UniqueMerger.h (hunk 3 above)
Comment (rappoccio) [reply CommonTools/UtilAlgos/interface/UniqueMerger.h:30]: We need it back, since we have the LogWarning when enabled by a user flag.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> **Failed tests: HeaderConsistency**

that's about the messageLogger include

**Comment by rappoccio** [comment]
OK, do we have to change anything? Is there a new way to include this?

**Comment by slava77** [comment]
> OK, do we have to change anything? Is there a new way to include this?

are you asking how to include header files in C++?

**Comment by rappoccio** [comment]
No, sorry, it looks like that line was somehow removed by mistake.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by santocch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/30251	2f7341c 
- code changes are in line with the PR description and the follow up review; based on visual inspection (the merger modified in this PR is not yet used in tested workflows)
- jenkins tests pass and comparisons with the baseline show no differences

**Comment by silviodonato** [comment]
+1

**Pull request merged**